packages/server/src/unstable-core-do-not-import/stream/jsonl.ts

```diff
@@ -266,20 +247,42 @@ function createBatchStreamProducer(opts: ProducerOptions) {
   for (const [key, item] of Object.entries(data)) {
     newHead[key] = encode(item, [key]);
   }
-  pending.activate();
 
-  return [newHead, stream.readable] as const;
+  yield newHead;
+  // Process all async iterables in parallel by racing their next values
+  try {
+    // Process while there are active iterators
+    while (set.size > 0) {
+      // Race all iterators to get the next value from any of them
+      const [entry, res] = await Unpromise.race(
+        Array.from(set).map(async (it) => [it, await it.nextPromise] as const),
+      );
+
+      yield res.value;
+
+      // Remove current iterator and re-add if not done
+      set.delete(entry);
+      if (!res.done) {
+        entry.nextPromise = entry.iterator.next();
+        set.add(entry);
+      }
+    }
+  } finally {
+    // Properly clean up any remaining iterators by calling return()
+    // Ensures resources are released if the loop exits early (e.g. due to error)
+    await Promise.all(Array.from(set).map((it) => it.iterator.return?.()));
+    set.clear();
+  }
```

Comment:
Iterating over sets is a slow operation. We use it frequently here. Since we don't add new entries after the initial `encode`s, a RingBuffer structure would be optimal. This should improve readability as we get rid of the false pretense of using sets while also improving performance.

This is too much to implement as a review feedback, but keep in mind as a possible refactor.

Reply:
Never used a ring buffer - wanna show how in a fork of this fork?